crates/commitlog/src/lib.rs

```diff
@@ -51,3 +64,250 @@ impl Default for Options {
+}
+
+impl<T> Commitlog<T> {
+    /// Open the log at root directory `root` with [`Options`].
+    ///
+    /// The root directory must already exist.
+    pub fn open(root: impl Into<PathBuf>, opts: Options) -> io::Result<Self> {
+        let inner = commitlog::Generic::open(repo::Fs::new(root), opts)?;
+
+        Ok(Self {
+            inner: RwLock::new(inner),
+        })
+    }
+
+    /// Sync all OS-buffered writes to disk.
+    ///
+    /// Note that this does **not** write outstanding records to disk.
+    /// Use [`Self::flush_and_sync`] or call [`Self::flush`] prior to this
+    /// method to ensure all data is on disk.
+    ///
+    /// Returns the maximum transaction offset which is considered durable after
+    /// this method returns successfully. The offset is `None` if the log hasn't
+    /// been flushed to disk yet.
+    pub fn sync(&self) -> io::Result<Option<u64>> {
+        let inner = self.inner.read().unwrap();
+        info!("sync commitlog");
+        inner.sync()?;
+
+        Ok(inner.max_committed_offset())
+    }
+
+    /// Write all outstanding transaction records to disk.
+    ///
+    /// Note that this does **not** force the OS to flush the data to the
+    /// filesystem. Use [`Self::flush_and_sync`] or call [`Self::sync`] after
+    /// this method to ensure all data is on disk.
+    ///
+    /// Returns the maximum transaction offset written to disk. The offset is
+    /// `None` if the log is empty and no data was pending to be flushed.
+    ///
+    /// Repeatedly calling this method may return the same value.
+    pub fn flush(&self) -> io::Result<Option<u64>> {
+        let mut inner = self.inner.write().unwrap();
+        info!("flush commitlog");
+        inner.commit()?;
+
+        Ok(inner.max_committed_offset())
+    }
+
+    /// Write all outstanding transaction records to disk and flush OS buffers.
+    ///
+    /// Equivalent to calling [`Self::flush`] followed by [`Self::sync`], but
+    /// without releasing the write lock in between.
+    pub fn flush_and_sync(&self) -> io::Result<Option<u64>> {
+        let mut inner = self.inner.write().unwrap();
+        info!("flush and sync commitlog");
+        inner.commit()?;
+        inner.sync()?;
+
+        Ok(inner.max_committed_offset())
+    }
+
+    /// Obtain an iterator which traverses the log from the start, yielding
+    /// [`Commit`]s.
+    ///
+    /// The returned iterator is not aware of segment rotation. That is, if a
+    /// new segment is created after this method returns, the iterator will not
+    /// traverse it.
+    ///
+    /// Commits appended to the log while it is being traversed are generally
+    /// visible to the iterator. Upon encountering [`io::ErrorKind::UnexpectedEof`],
+    /// however, a new iterator should be created using [`Self::commits_from`]
+    /// with the last transaction offset yielded.
+    pub fn commits(&self) -> impl Iterator<Item = Result<Commit, error::Traversal>> {
```

Comment:
slash-slash-todo: We're skipping bad commits in case the next commit is good. It's not obvious what to do when the iterator is exhausted and the last commit is bad -- there could be more coming at a later point in time. So for now, we'll yield the error.

This is fine in tooling, but requires special-casing when replaying the log to fold it into a datastore. Maybe a named iterator which can be turned into one which just ignores the error if it's the last item could be handy.

Reply:
I settled on yielding the error in iterators, but ignore it in folds (because the consumer of a fold doesn't have access to the traversal itself).